src/Block/Service/AbstractMenuBlockService.php

```diff
@@ -0,0 +1,246 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\BlockBundle\Block\Service;
+
+use Knp\Menu\ItemInterface;
+use Sonata\BlockBundle\Block\BlockContextInterface;
+use Sonata\BlockBundle\Form\Mapper\FormMapper;
+use Sonata\BlockBundle\Meta\Metadata;
+use Sonata\BlockBundle\Meta\MetadataInterface;
+use Sonata\BlockBundle\Model\BlockInterface;
+use Sonata\Form\Type\ImmutableArrayType;
+use Sonata\Form\Validator\ErrorElement;
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormTypeInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Twig\Environment;
+
+abstract class AbstractMenuBlockService extends AbstractBlockService implements EditableBlockService
+{
+    public function __construct(Environment $twig)
+    {
+        parent::__construct($twig);
+    }
+
+    public function execute(BlockContextInterface $blockContext, ?Response $response = null): Response
+    {
+        $template = $blockContext->getTemplate();
+        \assert(null !== $template);
+
+        $responseSettings = [
+            'menu' => $this->getMenu($blockContext),
+            'menu_options' => $this->getMenuOptions($blockContext->getSettings()),
+            'block' => $blockContext->getBlock(),
+            'context' => $blockContext,
+        ];
+
+        // NEXT_MAJOR: remove
+        if ('private' === $blockContext->getSetting('cache_policy')) {
+            return $this->renderPrivateResponse($template, $responseSettings, $response);
+        }
+
+        return $this->renderResponse($template, $responseSettings, $response);
+    }
+
+    public function configureCreateForm(FormMapper $form, BlockInterface $block): void
+    {
+        $this->configureEditForm($form, $block);
+    }
+
+    public function configureEditForm(FormMapper $form, BlockInterface $block): void
+    {
+        $form->add('settings', ImmutableArrayType::class, [
+            'keys' => $this->getFormSettingsKeys(),
+            'translation_domain' => 'SonataBlockBundle',
+        ]);
+    }
+
+    public function validate(ErrorElement $errorElement, BlockInterface $block): void
+    {
+    }
+
+    /**
+     * NEXT_MAJOR: Remove this method.
```

Comment:
This will require this: https://github.com/sonata-project/SonataBlockBundle/pull/1113 if it's fine for you @jordisala1991 @core23

Reply:
Can we separate those changes? I am not really sure TBH.

Reply:
The issue is the fact you need to provide a Metadata so far, and a metadata name then.

Does it make sens to provide a Metadata in an abstract class ? This would mean that every block extending it will have the same. (Unless they override it).

Reply:
You don't have to, just declare the method abstract and implement in the next class.

Reply:
i see 3 options:

1. declare the method normally here and make it abstract on 5.x (not sure how good the idea will be in terms of supporting multiple block bundle versions) to avoid BC break
2. split the abstract block in a class + trait (like we discussed on slack) to avoid BC break
3. Assume the bc break.

Reply:
In terms of avoiding BC breaks, imo the 2nd is the best, because it will allow to not have an editable block for breadcrumbs.

Reply:
I don't really like trait so wanted to avoid solution 2 and i'm more in favor of the solution 1.
This seems ok for the support of multiple block bundle version, you can override the method.

But still I wonder if the split of the interface is not a good idea. (but can be later)
You might edit those block without the need of a Metadata.